src/eventual/ptr.rs

```diff
@@ -0,0 +1,126 @@
+use by_address::ByAddress;
+use std::{
+    borrow::Borrow, cmp::Ordering, convert::AsRef, error::Error, fmt, hash::Hash, ops::Deref,
+    sync::Arc,
+};
+
+/// This type is a thin wrapper around T to enable cheap clone and comparisons.
+/// Internally it is an Arc that is compared by address instead of by the
+/// implementation of the pointed to value.
+///
+/// Additionally, Ptr implements Error where T: Error. This makes working with
+/// TryEventuals easier since many error types do not impl Value, but when
+/// wrapped in Ptr do.
+///
+/// One thing to be aware of is that because values are compared by address
+/// subscribers may be triggered unnecessarily in some contexts. If this is
+/// undesirable use Arc instead.
+///
+/// This type is not specificaly Eventual related, but is a useful pattern.
+#[repr(transparent)]
+#[derive(Debug, Default)]
+pub struct Ptr<T> {
+    inner: ByAddress<Arc<T>>,
+}
+
+impl<T> Ptr<T> {
+    #[inline]
+    pub fn new(wrapped: T) -> Self {
+        Self {
+            inner: ByAddress(Arc::new(wrapped)),
+        }
+    }
+}
+
+impl<T> Deref for Ptr<T> {
+    type Target = T;
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.inner.deref()
+    }
+}
+
+impl<T> Borrow<T> for Ptr<T> {
+    #[inline]
+    fn borrow(&self) -> &T {
+        self.inner.borrow()
+    }
+}
+
+impl<T> AsRef<T> for Ptr<T> {
+    #[inline]
+    fn as_ref(&self) -> &T {
+        self.inner.as_ref()
+    }
+}
+
+impl<T> Hash for Ptr<T> {
+    #[inline]
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.inner.hash(state)
```

Comment:
Never mind, the [docs](https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html) say that the default hashing algorithm is currently SipHash.

Reply:
Yes, all this does is feed the integral value to the Hasher. This API (feed data to a hasher) makes sense over the alternative (return a hash of data) because when used in context (eg: when there is a containing struct) it doesn't co-mingle the output and the algorithm, or require setting up the Hasher more than once. It was a surprisingly good choice to me after coming from C#/.NET which instead has an API that returns the hash.